pkg/nsx/services/firewall.go

```diff
@@ -201,45 +218,77 @@ func (service *SecurityPolicyService) buildBasicTags(obj *v1alpha1.SecurityPolic
 	return tags
 }
 
-func (service *SecurityPolicyService) updateTargetExpressions(obj *v1alpha1.SecurityPolicy, target *v1alpha1.SecurityPolicyTarget, group *model.Group, idx int) {
+func (service *SecurityPolicyService) updateTargetExpressions(obj *v1alpha1.SecurityPolicy, target *v1alpha1.SecurityPolicyTarget, group *model.Group, idx int) int {
 	service.appendOperatorIfNeeded(&group.Expression, "OR")
-	expressions := data.NewListValue()
-	expressionFrame := data.NewStructValue(
-		"",
-		map[string]data.DataValue{
-			"expressions":   expressions,
-			"resource_type": data.NewStringValue("NestedExpression"),
-		},
-	)
-	group.Expression = append(group.Expression, expressionFrame)
+	expressions := service.buildGroupExpression(&group.Expression)
 
 	clusterExpression := service.buildExpression(
 		"Condition", "SegmentPort",
 		fmt.Sprintf("%s|%s", util.TagScopeNCPCluster, service.getCluster()),
-		"Tag", "EQUALS",
+		"Tag", "EQUALS", "EQUALS",
 	)
 	expressions.Add(clusterExpression)
+	var err error = nil
+	var podSelectorCount, vMSelectorCount = 0, 0
+
 	if target.PodSelector != nil {
 		service.addOperatorIfNeeded(expressions, "AND")
 		// TODO: consider to use project_uid instead of project
 		nsExpression := service.buildExpression(
 			"Condition", "SegmentPort",
 			fmt.Sprintf("%s|%s", util.TagScopeNCPProject, obj.ObjectMeta.Namespace),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(nsExpression)
-		service.updatePortExpressions(target.PodSelector.MatchLabels, expressions)
+		service.updateExpressionsMatchLables(target.PodSelector.MatchLabels, "SegmentPort", expressions)
+		if len(target.PodSelector.MatchLabels) > 0 {
+			podSelectorCount = 1
```

Comment:
Do you mean `podSelectorCount += 1`?

Reply:
No, if only podSelector used, all labelSelectors for podSelector will produce one group. So podSelectorCount should be one.

Reply:
destinations:
        - podSelector:
            matchLabels:
              role: backend
            matchExpressions:
              - {key: k2, operator: NotIn, values: [b1,b2]}
              - {key: k2, operator: NotIn, values: [b1,b2,b3]}
              - 

      - podSelector:
            matchLabels:
              role: backend
            matchExpressions:
              - {key: k2, operator: NotIn, values: [b1,b2]}
              - {key: k2, operator: NotIn, values: [b1,b2,b3]}
   
The above podSelector will one produce one group criteria:
<img width="973" alt="image" src="https://user-images.githubusercontent.com/78414502/150450670-b2138910-c86c-4832-8d90-3ed553a55475.png">